File: modules/beep-widget/ios/BeepWidgetModule.swift
import ExpoModulesCore
import WidgetKit

public class BeepWidgetModule: Module {
    public func definition() -> ModuleDefinition {
        Name("BeepWidget")

        Function("updateWidgetData") { (data: String) in
            guard let userDefaults = UserDefaults(suiteName: "group.com.beepget.shared") else { return }
            userDefaults.set(data, forKey: "widget_data")
            userDefaults.synchronize()
            if #available(iOS 14.0, *) {
                WidgetCenter.shared.reloadAllTimelines()
            }
        }

        Function("reloadWidgets") {
            if #available(iOS 14.0, *) {
                WidgetCenter.shared.reloadAllTimelines()
            }
        }

        AsyncFunction("getWidgetData") { () -> String? in
            guard let userDefaults = UserDefaults(suiteName: "group.com.beepget.shared") else { return nil }
            return userDefaults.string(forKey: "widget_data")
        }
    }
}

File: modules/beep-widget/android/src/main/java/expo/modules/beepwidget/BeepWidgetModule.kt
package expo.modules.beepwidget

import android.content.Context
import androidx.glance.appwidget.updateAll
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class BeepWidgetModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("BeepWidget")

        Function("updateWidgetData") { data: String ->
            val context = appContext.reactContext ?: return@Function null
            BeepWidgetData.save(context, data)
            triggerWidgetUpdate(context)
            null
        }

        Function("reloadWidgets") {
            val context = appContext.reactContext ?: return@Function null
            triggerWidgetUpdate(context)
            null
        }

        AsyncFunction("getWidgetData") {
            val context = appContext.reactContext ?: return@AsyncFunction null
            BeepWidgetData.load(context)
        }
    }

    private fun triggerWidgetUpdate(context: Context) {
        val appContext = context.applicationContext
        CoroutineScope(Dispatchers.Default).launch {
            BeepWidgetSmall().updateAll(appContext)
            BeepWidgetMedium().updateAll(appContext)
        }
    }
}
